### PR TITLE
Clang format as a build target

### PR DESCRIPTION
**Pull request by jroweboy**
See https://github.com/yuzu-emu/yuzu/pull/117 for more details.

Please review it commit by commit. This also depends on https://github.com/citra-emu/ext-windows-bin/pull/15 for letting windows download a clang-format binary that they can use directly.

## What is all here

* Generates an optional build target when cmake can find the right version of clang-format. For MSVC, you can right click the clang-format project and have it format all source files. For Makefiles, you can run `make clang-format`.

* Each platform has a custom command that is run to gather all the files when the clang-format is run and not when the project is generated. This means that untracked files will also be formatted. 

* Bumps up the version to clang-format 6.0 which is the version installed when installing the MSVC plugin.

* For a bonus change, I added conditional options to automatically turn on `USE_BUNDLED_*` when MSVC is the toolset and the `ENABLE_*` option is ON. It always seemed silly to me that we make you click those check boxes ... After this is merged, I'll remove those steps from the wiki.

## If you would like to test this:

I've updated the instructions on the contributing guide with how to run clang format. Please try it out and let me know if anything is unclear.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3410)

### CODE REVIEW

**Comment by wwylele** [comment]
Idea LGTM, but for current state of citra, I have some concern about bumping the clang-format version:
 - current CI status is bad (Travis CI still cleaning up the queue, Appveyor canary build gets stuck in the queue...)
 - Some big PRs pending (multiplayer, texcache...)
 - Less people active for now due to yuzu
Due to this situation, the huge PR rebase required by an immediate clang-format version bump may be significantly delayed and poor reviewed, resulting a wave of corrupted PR.

My suggestion is
 - either delay this PR until we cleaned up some PR and back to the normal active development routine
 - or merge this PR, but remove the CI check for clang-format and don't do any reformat. After everything is settled down, bring the CI check back and reformat everything.

File: CMakeLists.txt [hunk 1]
```diff
@@ -275,6 +278,53 @@ if (UNIX OR MINGW)
...
 endif()
 
+# Setup a custom clang-format target (if clang-format can be found) that will run
+# against all the src files. This should be used before making a pull request.
+# =======================================================================
+
+set(CLANG_FORMAT_POSTFIX "-6.0")
+find_program(CLANG_FORMAT
+    NAMES clang-format${CLANG_FORMAT_POSTFIX}
+          clang-format
```
Comment (B3n30) [inline CMakeLists.txt]: I don't like that it falls back to `clang-format` without checking it's version if `clang-format-6.0` wasn't available

File: src/audio_core/hle/source.h [hunk 2]
```diff
@@ -5,8 +5,8 @@
 #pragma once
 
 #include <array>
-#include <queue>
 #include <vector>
+#include <queue>
```
Comment (B3n30) [inline src/audio_core/hle/source.h]: Why is this changed?

Comment (jroweboy) [reply src/audio_core/hle/source.h]: i... don't know actually. thats strange.

**Comment by B3n30** [comment]
Some notes: We have some regions in the code defined with clang-format off mostly to have nice lists. With the new clang-format version is could be those definitions of regions aren't necessary anymore. So they probably should get removed (if possible) within that PR. 

I also think our Contributing.md should get changed to account for that PR

**Comment by jroweboy** [comment]
@B3n30 I do update it https://github.com/citra-emu/citra/pull/3410/commits/52b3f00d79abe64c148c9567a83107b28a7e9c55 what exactly needs mentioned?

**Comment by adityaruplaha** [comment]
Is this conflicting with Multiplayer?

**Comment by BreadFish64** [comment]
@adityaruplaha this isn't even in Canary, don't worry about it

**Comment by wwylele** [comment]
I am gonna set a roadmap of dealing with this:
 - Merge #3071, #3073, #3281, #3440 and #3444 should be merged first
 - Then rebase this PR on master, reformat everything, and immediately merge.
 - Until this being merged, all PRs should still follow the old CI check (clang-format-3.9 style)
 - After this being merged, rebase and reformat all open PR
 - Any big PR, including incoming glvtx, should be reformat and open after this being merged.
